### PR TITLE
Drop support for Python 3.8

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
3.8 is EOL in late October 2024, and 3.13 final is scheduled for early October 2024. Aiming for new minor releases of astroid and pylint by then to drop support for 3.8 and add support for 3.13.

### CODE REVIEW

**Comment by jacobtylerwalls** [comment]
pre-commit.ci autofix

File: astroid/rebuilder.py [hunk 1]
```diff
@@ -80,10 +80,6 @@ def _get_doc(self, node: T_Doc) -> tuple[T_Doc, ast.Constant | ast.Str | None]:
                 ):
                     doc_ast_node = first_value
                     node.body = node.body[1:]
-                    # The ast parser of python < 3.8 sets col_offset of multi-line strings to -1
-                    # as it is unable to determine the value correctly. We reset this to None.
-                    if doc_ast_node.col_offset == -1:
-                        doc_ast_node.col_offset = None
```
Comment (Pierre-Sassoulas) [inline astroid/rebuilder.py:86]: Ho, this one is going to be a nice cleanup in pylint's testutil / functional !

File: astroid/rebuilder.py [hunk 2]
```diff
@@ -194,7 +190,7 @@ def visit_module(
             [self.visit(child, newnode) for child in node.body],
             doc_node=self.visit(doc_ast_node, newnode),
         )
-        if IS_PYPY and PY38:
+        if IS_PYPY:
             self._reset_end_lineno(newnode)
```
Comment (Pierre-Sassoulas) [inline astroid/rebuilder.py]: Shouldn't we remove the condition and everything in it here as it was an AND and can never be true now ?

File: astroid/rebuilder.py [hunk 3]
```diff
@@ -542,13 +528,12 @@ def visit_arguments(self, node: ast.arguments, parent: NodeNG) -> nodes.Argument
...
-            for arg in node.kwonlyargs:
-                if arg.annotation is not None:
-                    arg.end_lineno = arg.annotation.end_lineno
-                    arg.end_col_offset = arg.annotation.end_col_offset
+        # In Python 3.8 'end_lineno' and 'end_col_offset'
+        # for 'kwonlyargs' don't include the annotation.
+        for arg in node.kwonlyargs:
+            if arg.annotation is not None:
+                arg.end_lineno = arg.annotation.end_lineno
+                arg.end_col_offset = arg.annotation.end_col_offset
```
Comment (Pierre-Sassoulas) [inline astroid/rebuilder.py]:
```suggestion
```
AFAIU ?

File: tests/test_nodes_lineno.py [hunk 4]
```diff
@@ -8,12 +8,11 @@
...
 import astroid
 from astroid import builder, nodes
-from astroid.const import IS_PYPY, PY38, PY39_PLUS, PY310_PLUS, PY312_PLUS
+from astroid.const import IS_PYPY, PY310_PLUS, PY312_PLUS
 
 
 @pytest.mark.skipif(
-    not (PY38 and IS_PYPY),
-    reason="end_lineno and end_col_offset were added in PY38",
+    not IS_PYPY, reason="end_lineno and end_col_offset were added in PY38"
```
Comment (Pierre-Sassoulas) [inline tests/test_nodes_lineno.py]: Ditto, we should remove the whole test, it was skipped if not PY38 ?

File: astroid/nodes/scoped_nodes/scoped_nodes.py [hunk 5]
```diff
@@ -2007,7 +2007,7 @@ def fromlineno(self) -> int:
 
         Can also return 0 if the line can not be determined.
         """
-        if IS_PYPY and PY38 and not PYPY_7_3_11_PLUS:
+        if IS_PYPY:
```
Comment (Pierre-Sassoulas) [inline astroid/nodes/scoped_nodes/scoped_nodes.py]: Should remove everything as it's never going to be true again ?

**Review by Pierre-Sassoulas — COMMENTED** [review]
Releasing the review before completion because I don't want to comment n time the same thing, if I'm misguided in saying that every condition that has ``PY38 and`` something else can be removed entirely.

File: ChangeLog [hunk 6]
```diff
@@ -7,21 +7,15 @@ What's New in astroid 3.3.0?
...
 
-* Fix ``RecursionError`` in ``infer_call_result()`` for certain ``__call__`` methods.
-
-  Closes pylint-dev/pylint#9139
-
-* Add ``AstroidManager.prefer_stubs`` attribute to control the astroid 3.2.0 feature that prefers stubs.
-
-  Refs pylint-dev/#9626
-  Refs pylint-dev/#9623
-
```
[the remark this answers is not in the document]
Comment (jacobtylerwalls) [reply ChangeLog:24]: Duplicate snuck in somehow. See 3.2.1 below.

**Comment by jacobtylerwalls** [comment]
Yeah, I just caught that myself. Sorry!

File: ChangeLog (hunk 6 above)
[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply ChangeLog:24]: Ha, the one thing to look for when merging back maintenance in main, I missed it during review.

File: astroid/brain/brain_hashlib.py [hunk 7]
```diff
@@ -4,12 +4,11 @@
 
 from astroid.brain.helpers import register_module_extender
 from astroid.builder import parse
-from astroid.const import PY39_PLUS
 from astroid.manager import AstroidManager
 
 
 def _hashlib_transform():
-    maybe_usedforsecurity = ", usedforsecurity=True" if PY39_PLUS else ""
+    maybe_usedforsecurity = ", usedforsecurity=True"
```
Comment (cdce8p) [inline astroid/brain/brain_hashlib.py:11]: Should this just be inlined now?

File: astroid/brain/brain_subprocess.py [hunk 8]
```diff
@@ -89,8 +88,7 @@ def kill(self):
         {ctx_manager}
        """
     )
-    if PY39_PLUS:
-        code += """
+    code += """
     @classmethod
     def __class_getitem__(cls, item):
         pass
```
Comment (cdce8p) [inline astroid/brain/brain_subprocess.py:94]: The `__class_getitem__` method can be added to the previous code block.

File: astroid/brain/brain_typing.py [hunk 9]
```diff
@@ -472,14 +454,9 @@ def register(manager: AstroidManager) -> None:
...
     )
 
-    if PY39_PLUS:
-        manager.register_transform(
-            FunctionDef, inference_tip(infer_typedDict), _looks_like_typedDict
-        )
-    else:
-        manager.register_transform(
-            ClassDef, inference_tip(infer_old_typedDict), _looks_like_typedDict
-        )
```
Comment (cdce8p) [inline astroid/brain/brain_typing.py:482]: I believe `infer_old_typedDict` is now unused and can be removed as well.

File: astroid/rebuilder.py [hunk 10]
```diff
@@ -1733,9 +1672,6 @@ def _visit_with(
         parent: NodeNG,
     ) -> _WithT:
         col_offset = node.col_offset
```
Comment (cdce8p) [inline astroid/rebuilder.py:1674]: `col_offset` can be inlined now a few lines below in the `cls` call.

File: astroid/rebuilder.py [hunk 11]
```diff
@@ -1048,9 +1005,6 @@ def _visit_for(
     ) -> _ForT:
         """Visit a For node by returning a fresh instance of it."""
         col_offset = node.col_offset
```
Comment (cdce8p) [inline astroid/rebuilder.py:1007]: Same here

File: tests/test_nodes_lineno.py [hunk 12]
```diff
@@ -200,13 +165,9 @@ def test_end_lineno_call() -> None:
...
 
         # fmt: off
-        if PY39_PLUS:
-            # 'lineno' and 'col_offset' information only added in Python 3.9
-            assert (c1.keywords[0].lineno, c1.keywords[0].col_offset) == (1, 11)
-            assert (c1.keywords[0].end_lineno, c1.keywords[0].end_col_offset) == (1, 21)
-        else:
-            assert (c1.keywords[0].lineno, c1.keywords[0].col_offset) == (None, None)
-            assert (c1.keywords[0].end_lineno, c1.keywords[0].end_col_offset) == (None, None)
+        # 'lineno' and 'col_offset' information only added in Python 3.9
```
Comment (cdce8p) [inline tests/test_nodes_lineno.py:168]:
```suggestion
```
Can probably be removed now

File: tests/test_nodes_lineno.py [hunk 13]
```diff
@@ -842,13 +803,9 @@ def test_end_lineno_subscript() -> None:
...
         assert (s3.lineno, s3.col_offset) == (3, 0)
         assert (s3.end_lineno, s3.end_col_offset) == (3, 11)
-        if PY39_PLUS:
-            # 'lineno' and 'col_offset' information only added in Python 3.9
-            assert (s3.slice.lineno, s3.slice.col_offset) == (3, 4)
-            assert (s3.slice.end_lineno, s3.slice.end_col_offset) == (3, 10)
-        else:
-            assert (s3.slice.lineno, s3.slice.col_offset) == (None, None)
-            assert (s3.slice.end_lineno, s3.slice.end_col_offset) == (None, None)
+        # 'lineno' and 'col_offset' information only added in Python 3.9
```
Comment (cdce8p) [inline tests/test_nodes_lineno.py:806]:
```suggestion
```
Same

File: tests/test_nodes_lineno.py [hunk 14]
```diff
@@ -1246,13 +1191,9 @@ class X(Parent, var=42):
...
         assert (c1.bases[0].lineno, c1.bases[0].col_offset) == (3, 8)
         assert (c1.bases[0].end_lineno, c1.bases[0].end_col_offset) == (3, 14)
-        if PY39_PLUS:
-            # 'lineno' and 'col_offset' information only added in Python 3.9
-            assert (c1.keywords[0].lineno, c1.keywords[0].col_offset) == (3, 16)
-            assert (c1.keywords[0].end_lineno, c1.keywords[0].end_col_offset) == (3, 22)
-        else:
-            assert (c1.keywords[0].lineno, c1.keywords[0].col_offset) == (None, None)
-            assert (c1.keywords[0].end_lineno, c1.keywords[0].end_col_offset) == (None, None)
+        # 'lineno' and 'col_offset' information only added in Python 3.9
```
Comment (cdce8p) [inline tests/test_nodes_lineno.py:1194]:
```suggestion
```

**Review by cdce8p — COMMENTED** [review]
I'd probably have done that in multiple steps / PRs, like
- Update CI + do code changes to remove 3.8 special cases
- Update tests
- Update typing + ruff and pylint to 3.9.0

Might be easier to follow when reviewing it later, but this works the same.

--
In general, I do support removing support for 3.8. We could even do it fairly soon IMO. Maybe after the PEP 695 fix is merged and one last bugfix release created?

For pylint, we should just make sure to bump the minor version and update `requires-python` when we pull in the new astroid release. The code changes to remove 3.8 can follow later as well.

**Review by cdce8p — APPROVED** [review]
LGTM! Thanks @jacobtylerwalls 👍🏻

File: astroid/_backport_stdlib_names.py [hunk 15]
```diff
@@ -346,11 +346,7 @@
...
 
-if sys.version_info[:2] == (3, 7):
-    stdlib_module_names = PY_3_7
-elif sys.version_info[:2] == (3, 8):
-    stdlib_module_names = PY_3_8
-elif sys.version_info[:2] == (3, 9):
+if sys.version_info[:2] == (3, 9):
     stdlib_module_names = PY_3_9
 else:
     raise AssertionError("This module is only intended as a backport for Python <= 3.9")
```
Comment (Pierre-Sassoulas) [inline astroid/_backport_stdlib_names.py:352]: Can't we remove the whole module?

File: astroid/modutils.py [hunk 16]
```diff
@@ -74,20 +73,6 @@
         except AttributeError:
             pass
 
-if IS_PYPY and sys.version_info < (3, 8):
```
Comment (Pierre-Sassoulas) [inline astroid/modutils.py:77]: Great catch, look like we forgot to remove when dropping 3.7

**Review by Pierre-Sassoulas — COMMENTED** [review]
byebye400lines

... So much joy.

File: astroid/_backport_stdlib_names.py (hunk 15 above)
Comment (Pierre-Sassoulas) [reply astroid/_backport_stdlib_names.py:352]: Ha, it's ``<=``my bad nvm

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**